### PR TITLE
Make generated consumers support message keys

### PR DESCRIPTION
**Pull request by jerrypnz**

### CODE REVIEW

File: lib/src/test/resources/kafka-consumer.txt [hunk 1]
```diff
@@ -114,52 +114,66 @@ package test.apidoc.apidoctest.v0.kafka {
     def processBatchThenCommit(
       processor: Map[String, Seq[KafkaMember]] ⇒ scala.util.Try[Map[String, Seq[KafkaMember]]],
       batchSize: Int = 1
-    ): scala.util.Try[Map[String, Seq[KafkaMember]]] = {
+    ): scala.util.Try[Map[String, Seq[KafkaMember]]] =
+      doProcess[KafkaMember] { message ⇒
+        Option(message.message).map(Json.parse(_).as[KafkaMember])
+      }(processor, batchSize)
+
+    def processBatchWithKeysThenCommit(
```
Comment (jackhopner) [inline lib/src/test/resources/kafka-consumer.txt]: Maybe some documentation explaining the signature of this method and its usage.

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 2]
```diff
@@ -163,49 +163,63 @@ package ${ssd.namespaces.base}.kafka {
     def processBatchThenCommit(
       processor: Map[String, Seq[${className}]] ⇒ scala.util.Try[Map[String, Seq[${className}]]],
       batchSize: Int = 1
-    ): scala.util.Try[Map[String, Seq[${className}]]] = {
+    ): scala.util.Try[Map[String, Seq[${className}]]] =
+      doProcess[${className}] { message ⇒
+        Option(message.message).map(Json.parse(_).as[${className}])
+      }(processor, batchSize)
+
+    def processBatchWithKeysThenCommit(
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:182]: Maybe some documentation explaining the signature of this method and its usage.

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 3]
```diff
@@ -36,14 +36,14 @@ object KafkaConsumer extends CodeGenerator {
     val play2Json = Play2JsonExtended(ssd).generate()
 
     val header = addHeader match {
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: Would this be simpler as an if/else?

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 4]
```diff
@@ -163,49 +163,63 @@ package ${ssd.namespaces.base}.kafka {
...
       fetchBatch(batchSize, Map.empty) match {
-        case scala.util.Success(messages) =>
-          processor(messages) map { allMessages =>
+        case scala.util.Success(messages) ⇒
+          processor(messages) map { allMessages ⇒
             consumer.commitOffsets(true)
             allMessages
           }
-        case scala.util.Failure(ex) => scala.util.Failure(ex)
+        case scala.util.Failure(ex) ⇒ scala.util.Failure(ex)
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: should be on new line after =>?

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 5]
```diff
@@ -163,49 +163,63 @@ package ${ssd.namespaces.base}.kafka {
...
-                  messages + (tenant -> newSeq)
-                } else {
-                  messages
-                }
+                messages + (tenant → newSeq)
+              } getOrElse messages
 
               fetchBatch(remainingInBatch - 1, newMessages)
-            case scala.util.Failure(ex) => ex match {
+            case scala.util.Failure(ex) ⇒ ex match {
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: match should be on new line?

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 6]
```diff
@@ -163,49 +163,63 @@ package ${ssd.namespaces.base}.kafka {
...
+      processor: Map[String, Seq[(String, Option[${className}])]] ⇒ scala.util.Try[Map[String, Seq[(String, Option[${className}])]]],
+      batchSize: Int = 1
+    ): scala.util.Try[Map[String, Seq[(String, Option[${className}])]]] =
+      doProcess[(String,  Option[${className}])] { message ⇒
+        Some(
+          message.key → Option(message.message).map(Json.parse(_).as[${className}])
+        )
+      }(processor, batchSize)
+
+    def doProcess[T](
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:192]: Does this need to be generic if we're generating it?

File: scala-generator/src/main/scala/models/KafkaConsumer.scala [hunk 7]
```diff
@@ -163,49 +163,63 @@ package ${ssd.namespaces.base}.kafka {
...
+      converter: MessageAndMetadata[String, String] ⇒ Option[T]
+    )(
+      processor: Map[String, Seq[T]] ⇒ scala.util.Try[Map[String, Seq[T]]],
+      batchSize: Int = 1
+    ): scala.util.Try[Map[String, Seq[T]]] = {
       @tailrec
-      def fetchBatch(remainingInBatch: Int, messages: Map[String, Seq[${className}]]): scala.util.Try[Map[String, Seq[${className}]]] ={
+      def fetchBatch(remainingInBatch: Int, messages: Map[String, Seq[T]]): scala.util.Try[Map[String, Seq[T]]] ={
```
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:199]: generics required?

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 6 above)
Comment (jerrypnz) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:192]: It needs to be as it's used by both `processBatchThenCommit` and `processBatchWithKeysThenCommit` and they have different types of messages.

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 7 above)
Comment (jerrypnz) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:199]: The same reason as above.

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 4 above)
Comment (jerrypnz) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: Done

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 5 above)
Comment (jerrypnz) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: Done

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 6 above)
Comment (jackhopner) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala:192]: Sorry missed that, the type becomes quite complex with keys then.

File: scala-generator/src/main/scala/models/KafkaConsumer.scala (hunk 3 above)
Comment (jerrypnz) [inline scala-generator/src/main/scala/models/KafkaConsumer.scala]: Done

File: lib/src/test/resources/kafka-consumer.txt [hunk 8]
```diff
@@ -114,52 +114,66 @@ package test.apidoc.apidoctest.v0.kafka {
     def processBatchThenCommit(
```
Comment (jackhopner) [inline lib/src/test/resources/kafka-consumer.txt:117]: Don't think this test will pass currently?

**Comment by jackhopner** [comment]
lgtm other than the comment on the test.

File: lib/src/test/resources/kafka-consumer.txt (hunk 8 above)
Comment (jerrypnz) [inline lib/src/test/resources/kafka-consumer.txt:117]: Oops missed this one. Fixed.

**Pull request merged**